_source/_posts/2023-07-27-spring-graphql-react.md

```diff
@@ -0,0 +1,1794 @@
+  --name "React client for GraphQL" \
+  --description "SPA React client for a Spring GraphQL API" \
+  --type spa \
+  --callbacks http://localhost:3000/callback \
+  --logout-urls http://localhost:3000 \
+  --reveal-secrets
+```
+
+Copy the Auth0 domain and the client ID, and update the `src/.env.local` adding the following properties (add the new variables to the example file, too):
+
+__.env.local__
+```shell
+NEXT_PUBLIC_AUTH0_DOMAIN={yourAuth0Domain}
+NEXT_PUBLIC_AUTH0_CLIENT_ID={clientId}
+NEXT_PUBLIC_AUTH0_CALLBACK_URL=http://localhost:3000/callback
+NEXT_PUBLIC_AUTH0_AUDIENCE=https://{yourAuth0Domain}/api/v2/
+```
+
+For handling the Auth0 post-login behavior, you need to add the page `src/app/callback/page.tsx` with the following content.
+
+__callback/page.tsx__
+```tsx
+import Loader from "@/components/loader/Loader";
+
+const Page = () => {
+  return <Loader/>
+};
+
+export default Page;
+```
+
+For this example, the callback page will render empty.
+
+Add the `@auth0/auth0-react` dependency to the project:
+
+```shell
+npm install @auth0/auth0-react
+```
+
+Create the component `Auth0ProviderWithNavigate` in the folder `src/components/authentication` with the following content:
+
+{% raw %}
+__Auth0ProviderWithNavigate.tsx__
+```tsx
+import { AppState, Auth0Provider } from "@auth0/auth0-react";
+import { useRouter } from "next/navigation";
+import React from "react";
+
+const Auth0ProviderWithNavigate = (props: { children: React.ReactNode }) => {
+  const router = useRouter();
+
+  const domain = process.env.NEXT_PUBLIC_AUTH0_DOMAIN || "";
+  const clientId = process.env.NEXT_PUBLIC_AUTH0_CLIENT_ID || "";
+  const redirectUri = process.env.NEXT_PUBLIC_AUTH0_CALLBACK_URL || "";
+  const audience = process.env.NEXT_PUBLIC_AUTH0_AUDIENCE || "";
+
+  const onRedirectCallback = (appState?: AppState) => {
+    router.push(appState?.returnTo || window.location.pathname);
+  };
+
+  if (!(domain && clientId && redirectUri)) {
+    return null;
+  }
+
+  return (
+    <Auth0Provider
+      domain={domain}
+      clientId={clientId}
+      authorizationParams={{
+        audience: audience,
+        redirect_uri: redirectUri,
+      }}
+      useRefreshTokens={true}
+      onRedirectCallback={onRedirectCallback}
+    >
+      <>{props.children}</>
+    </Auth0Provider>
+  );
+};
+
+export default Auth0ProviderWithNavigate;
+```
+{% endraw %}
+
+The component `Auth0ProviderWithNavigate` wraps the children component with `Auth0Provider`, the provider of the Auth0 context, remembering the requested URL for redirection after login.
+Use the component in the `WideLayout` component:
+
+{% raw %}
+__WideLayout.tsx__
+```tsx
+const WideLayout = (props: { children: React.ReactNode }) => {
+  return (
+    <ThemeProvider theme={theme}>
+      <Auth0ProviderWithNavigate>
+        <Container maxWidth="lg" sx={{ mt: 4 }}>
+          {props.children}
+        </Container>
+      </Auth0ProviderWithNavigate>
+    </ThemeProvider>
+  );
+};
+```
```

Comment:
I had to add the following import for this to compile:

```
import Auth0ProviderWithNavigate from '@/components/authentication/Auth0ProviderWithNavigate';
```

Reply:
I updated the block showing the whole file.